File: warm_up/counting_valleys.py
#!/bin/python3

import math
import os
import random
import re
import sys

from functools import reduce


def countingValleys(steps, path):
    def f(height_valleys, direction):
        height, valleys = height_valleys
        if direction == 'D':
            return (height - 1, valleys)
        elif direction == 'U':
            height += 1
            valleys = valleys + 1 if height == 0 else valleys
            return (height, valleys)
        else:
            raise

    return reduce(f, path, (0, 0))[1]


if __name__ == '__main__':
    steps = int(input().strip())
    path = input()
    result = countingValleys(steps, path)

    print(str(result))

File: warm_up/counting_valleys.ml
open! Core

let command =
  let counting_valleys _steps path =
    String.fold path ~init:(0, 0) ~f:(fun (height, valleys) direction ->
        match direction with
        | 'D' -> height - 1, valleys
        | 'U' ->
          let height = height + 1 in
          let valleys = if height = 0 then valleys + 1 else valleys in
          height, valleys
        | _ -> failwith "Bad Input")
    |> snd
  in
  Common.Command_line.create () ~f:(fun () ->
      print_endline (counting_valleys 8 "DDUUDUDUDUDUDUUUUDDDD" |> Int.to_string))
;;
